File: trunk/simulations-java/simulations/beers-law-lab/src/edu/colorado/phet/beerslawlab/concentration/model/Shaker.java
// Copyright 2002-2012, University of Colorado
package edu.colorado.phet.beerslawlab.concentration.model;

import edu.colorado.phet.beerslawlab.common.model.Movable;
import edu.colorado.phet.common.phetcommon.math.Vector2D;
import edu.colorado.phet.common.phetcommon.model.property.Property;
import edu.colorado.phet.common.phetcommon.util.RichSimpleObserver;
import edu.colorado.phet.common.phetcommon.util.SimpleObserver;
import edu.umd.cs.piccolo.util.PBounds;

/**
 * Model of the shaker, contains solute in solid form.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 */
public class Shaker extends Movable {

    public final double orientation; // rotation angle, in radians
    public final Property<Solute> solute;
    public final Property<Boolean> visible;
    public final Property<Boolean> empty;
    private final double maxDispensingRate; // mol/sec
    private final Property<Double> dispensingRate; // mol/sec
    private Vector2D previousLocation;

    public Shaker( Vector2D location, double orientation, PBounds dragBounds, Property<Solute> solute, double maxDispensingRate ) {
        super( location, dragBounds );
        assert ( dragBounds.contains( location.toPoint2D() ) );

        this.orientation = orientation;
        this.solute = solute;
        this.visible = new Property<Boolean>( true );
        this.empty = new Property<Boolean>( false );
        this.maxDispensingRate = maxDispensingRate;
        this.dispensingRate = new Property<Double>( 0d );
        this.previousLocation = location;

        // set the dispensing rate to zero when the shaker becomes empty or invisible
        RichSimpleObserver rateObserver = new RichSimpleObserver() {
            public void update() {
                if ( empty.get() || !visible.get() ) {
                    dispensingRate.set( 0d );
                }
            }
        };
        rateObserver.observe( empty, visible );
    }

    public double getDispensingRate() {
        return dispensingRate.get();
    }

    public void addDispensingRateObserver( SimpleObserver observer ) {
        dispensingRate.addObserver( observer );
    }

    /*
     * This is called whenever the simulation clock ticks.
     * Sets the dispensing rate if the shaker is moving.
     */
    public void stepInTime() {
        if ( visible.get() && !empty.get() ) {
            if ( previousLocation.equals( location.get() ) ) {
                this.dispensingRate.set( 0d ); // shaker is not moving, don't dispense anything
            }
            else {
                this.dispensingRate.set( maxDispensingRate ); // this seems to work fine
            }
        }
        previousLocation = new Vector2D( getX(), getY() );
    }

    @Override public void reset() {
        super.reset();
        dispensingRate.reset();
        previousLocation = new Vector2D( getX(), getY() ); // to prevent shaker from dispensing solute when its location is reset
    }
}

File: simulations-html/simulations/beers-law-lab/js/concentration/model/Shaker.js
// Copyright 2002-2013, University of Colorado

/**
 * Model of a shaker, contains solute in solid form.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 */
define( [
            'phetcommon/math/Point2D',
            'phetcommon/model/property/Property',
            'common/model/Inheritance',
            'common/model/Movable'
        ],
        function ( Point2D, Property, Inheritance, Movable ) {

            /**
             * Constructor
             * @param {Point2D} location
             * @param {Number} orientation in radians
             * @param {Rectangle} dragBounds
             * @param {Property<Solute>} soluteProperty
             * @param {Number} maxDispensingRate
             * @constructor
             */
            function Shaker( location, dragBounds, orientation, soluteProperty, maxDispensingRate ) {

                Movable.call( this, location, dragBounds ); // constructor stealing

                this.orientation = orientation;
                this.soluteProperty = soluteProperty;
                this.maxDispensingRate = maxDispensingRate;

                this.visibleProperty = new Property( true );
                this.emptyProperty = new Property( false );
                this.dispensingRateProperty = new Property( 0 );
                this.previousLocation = location;

                // set the dispensing rate to zero when the shaker becomes empty or invisible
                var that = this;
                var observer = function() {
                    if ( that.emptyProperty.get() || !that.visibleProperty.get() ) {
                        that.dispensingRateProperty.set( 0 );
                    }
                };
                this.emptyProperty.addObserver( observer );
                this.visibleProperty.addObserver( observer );
            }

            Inheritance.inheritPrototype( Shaker, Movable );

            // Resets the shaker to its initial state.
            Shaker.prototype.resetSuper = Movable.prototype.reset; //TODO awful hack for calling super.reset
            Shaker.prototype.reset = function () {
                this.resetSuper();
                this.dispensingRateProperty.reset();
                this.previousLocation = this.locationProperty.get(); // to prevent shaker from dispensing solute when its location is reset
            }

            /*
             * This is called whenever the simulation clock ticks.
             * Sets the dispensing rate if the shaker is moving.
             */
            Shaker.prototype.tick = function () {
                if ( this.visibleProperty.get() && !this.emptyProperty.get() ) {
                    if ( this.previousLocation.equals( this.locationProperty.get() ) ) {
                        this.dispensingRateProperty.set( 0 ); // shaker is not moving, don't dispense anything
                    }
                    else {
                        this.dispensingRateProperty.set( this.maxDispensingRate ); // this seems to work fine
                    }
                }
                this.previousLocation = this.locationProperty.get();
            }

            return Shaker;
        } );